llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -3377,7 +3377,9 @@ X86TargetLowering::getJumpConditionMergingParams(Instruction::BinaryOps Opc,
                                                  const Value *Lhs,
                                                  const Value *Rhs) const {
   using namespace llvm::PatternMatch;
-  int BaseCost = BrMergingBaseCostThresh.getValue();
+  // Disable condition merging when CCMP is available b/c we can eliminate
+  // branches in a more efficient way.
+  int BaseCost = Subtarget.hasCCMP() ? -1 : BrMergingBaseCostThresh.getValue();
```

Comment:
Would have figured it would be easier to just lower `(and/or setcc, setcc)` as `cmpp` than essentially restitching the BBs at the MBB level (which seems to have some extra restricts IIUC `canConvert` correctly).

Reply:
Thinking

Reply:
Also note, if we are going to merge the BBs anyways, probably best to do as soon as possible so we have more context during DAG lowering.

Reply:
@goldsteinn We can't lower `(and/or setcc, setcc)` as `ccmp` b/c CCMP has a CMP semantic. Instead, what we need to need to do is sth like

1.  Combine the DAG
```
  // sub(and(setcc(cc1, ...), setcc(cc2, sub (X, Y))), 1)
  // brcond ne
  //
  //  ->
  //
  //  ccmp(X, Y, cc1, cc2)
  //  brcond cc2
  //
  //  if only flag has users.
```
2. Define fragment for `ccmp`
3. add pattern for CCMP using the fragment

I drafted a patch locally and found that it's more complex than the MIR pass.

IIUC, some checks in `canConvert` are conservative and not real restrictions, it's borrowed from AArch64ConditionalCompares.cpp, which was designed to run
immediately before the early if-conversion pass.

Reply:
I will upload the DAGCombine + pattern match solution too for comparison.

Reply:
> @goldsteinn We can't lower `(and/or setcc, setcc)` as `ccmp` b/c CCMP has a CMP semantic. Instead, what we need to need to do is sth like
> 
> 1. Combine the DAG
> 
> ```
>   // sub(and(setcc(cc1, ...), setcc(cc2, sub (X, Y))), 1)
>   // brcond ne
>   //
>   //  ->
>   //
>   //  ccmp(X, Y, cc1, cc2)
>   //  brcond cc2
>   //
>   //  if only flag has users.
> ```
> 
> 2. Define fragment for `ccmp`
> 3. add pattern for CCMP using the fragment
> 
> I drafted a patch locally and found that it's not simpler than the MIR pass.
> 
Is that #91747?
If so I disagree that it is not simpler... re-patching BBs in the MachineInst phase
is definitely a large source of the complexity of this patch.

> IIUC, some checks in `canConvert` are conservative and not real restrictions. It should not have more restriction than DAGCombine + pattern match approach. The function is borrowed from AArch64ConditionalCompares.cpp,
> 
> > probably best to do as soon as possible so we have more context during DAG lowering
> 
> Which context you think we need? AArch64ConditionalCompares.cpp was designed to run immediately before the early if-conversion pass.

What I mean is current DAGCombiner runs on BBs, so if are going to be merging BBs, doing so earlier
will improve the rest of lowering.

Based on the two missing cases in #91747, can they not be fixed on in simplifyCFG?

Reply:
> Is that https://github.com/llvm/llvm-project/pull/91747?
Yes.

> Based on the two missing cases in https://github.com/llvm/llvm-project/pull/91747, can they not be fixed on in simplifyCFG?

For the case that two compares are not in same BB, maybe we can fix it simplifyCFG.
For the case that DAGCombine only revisit the direct users after simplifying a node, we can only expect #91772 or similar PRs. TBH, I don't known when it can be ready.